rfcs/2021-12-20-10517-llvm-backend-for-vrl.md

```diff
@@ -0,0 +1,560 @@
+
+Statically linking LLVM to the Vector binary adds roughly 9MB, additionally to precompiled bitcode that needs to be included with the binary. If this is a concern, we can consider shipping binaries with the LLVM feature disabled.
+
+While LLVM is a highly used framework within the industry, working with it requires rather specialized knowledge about compiler construction. However, there exists plenty of publicly accessible material for code generation using LLVM, some of which I linked to [above](#introduction-to-llvm). In addition, there should be great focus to document this part of the code base extraordinarily well.
+
+## Prior Art
+
+_TODO._
+
+## Alternatives
+
+### Compile to Rust
+
+Using Rust as a compilation target would require us to
+
+- ship a Rust compiler and its libraries
+- ship Vector source code and its dependent crates
+
+which would be hundreds of MB and therefore infeasible.
+
+### Compile to C
+
+Using C as a compilation target would require us to
+
+- ship a C compiler
+
+while
+
+- not having any better safety guarantees
+- not being able to inline functions and therefore miss optimization potential
+
+and therefore not provide any significant benefits over using LLVM directly.
+
+### Compile to WebAssembly
+
+Using WebAssembly as a compilation target would require us to
+
+- ship a Wasm runtime
+- copy data in and out of WebAssembly or use `mmap`ing techniques which would constrain in which memory regions event data must reside
+- likely slower execution speed because of higher abstraction level and semantics that allow untrusted code to execute safely
+- precompile WebAssembly bitcode or miss optimization potential
+
+additionally to these drawbacks, we just dropped support for the WebAssembly transform.
+
+### Compile to Bitcode
+
+As mentioned in the context, we are currently moving forward with a VM for VRL. Compared to the current execution model and an LLVM-based approach, the VM provides a middle ground for execution speed, memory safety and sophistication.
+
+Weighing the benefits depends on the real world performance of both approaches.
+
+## Plan Of Attack
+
+Incremental steps to execute this change. These will be converted to issues after the RFC is approved:
+
+- Submit a PR with spike-level code _roughly_ demonstrating the change: [#10442](https://github.com/vectordotdev/vector/pull/10442).
+- Extract a core library from VRL for exposing its types with minimal dependencies, necessary to reduce size of the precompiled bitcode.
+- Refine code generation by taking into account type information.
+- Add unit tests for each expression in isolation.
+- Add fuzz testing infrastructure that compares results of all three execution modes.
```

Comment:
When we say infrastructure here the scope is limited to having tests available that can run or will also include a CI component? My preference would be for tests-only.

Reply:
Oh, my wording was a bit unfortunate, I agree there. The "infrastructure" was supposed to refer to "testing code" only

Reply:
Adressed in a1e95636c58e7229fda659db97858dfd2c05f757.